server/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java

```diff
@@ -4950,6 +4988,260 @@ public void testKeepMinRetainedSeqNoByMergePolicy() throws IOException {
+                 newMergePolicy(), null, null, globalCheckpoint::get))) {
+            engine.index(indexForDoc(createParsedDoc("1", null)));
+            engine.index(indexForDoc(createParsedDoc("1", null)));
+            engine.refresh("test");
+            engine.delete(new Engine.Delete("test", "1", newUid("1"), primaryTerm.get()));
+            engine.index(indexForDoc(createParsedDoc("1", null)));
+            engine.delete(new Engine.Delete("test", "1", newUid("1"), primaryTerm.get()));
+            engine.refresh("test");
+            List<Tuple<Long, Long>> seqNos = readUpdatedBySeqNos.apply(engine);
+            // index#0 -> index#1 -> delete#2 -> index#3 -> delete#4
+            assertThat(seqNos, containsInAnyOrder(
+                Tuple.tuple(0L, 1L),   // index#0 -> index#1 -> refresh
+                Tuple.tuple(1L, 4L),   // index#1 -> delete#2 -> index#3 -> delete#4
+                Tuple.tuple(2L, null), // the first delete tombstone without updated_by_seqno
+                Tuple.tuple(2L, 3L),   // the second delete tombstone with updated_by_seqno
+                Tuple.tuple(3L, 4L),   // index#3 -> delete#4
+                Tuple.tuple(4L, null)  // still alive
+            ));
+        }
+    }
+
+    @TestLogging("_ROOT:DEBUG")
+    public void testKeepDocsForRollback() throws Exception {
+        IOUtils.close(engine, store);
+        threadPool = spy(threadPool);
+        AtomicLong clockTime = new AtomicLong();
+        when(threadPool.relativeTimeInMillis()).thenAnswer(i -> clockTime.incrementAndGet());
+        List<Engine.Operation> operations = new ArrayList<>();
+        int numOps = scaledRandomIntBetween(10, 200);
+        for (int seqNo = 0; seqNo < numOps; seqNo++) {
+            String id = Integer.toString(between(1, 5));
+            if (randomBoolean()) {
+                ParsedDocument parseDoc = createParsedDoc(id, null);
+                operations.add(new Engine.Index(newUid(parseDoc), parseDoc, seqNo, primaryTerm.get(), 1, null,
+                    REPLICA, between(1, 10), IndexRequest.UNSET_AUTO_GENERATED_TIMESTAMP, randomBoolean()));
+            } else {
+                operations.add(new Engine.Delete("test", id, newUid(id), seqNo, primaryTerm.get(), 1, null,
+                    REPLICA, between(1, 10)));
+            }
+        }
+        Settings.Builder settings = Settings.builder()
+            .put(defaultSettings.getSettings())
+            .put(IndexSettings.INDEX_GC_DELETES_SETTING.getKey(), TimeValue.timeValueMillis(between(0, 100)).getStringRep())
+            .put(IndexSettings.INDEX_SOFT_DELETES_SETTING.getKey(), true);
+        final IndexMetaData indexMetaData = IndexMetaData.builder(defaultSettings.getIndexMetaData()).settings(settings).build();
+        final IndexSettings indexSettings = IndexSettingsModule.newIndexSettings(indexMetaData);
+        realisticShuffleOperations(operations);
```

Comment:
why be realistic here? Pure random is a better test, no?

Reply:
I ran several hundred iterations with the pure random but no failure. The reason is the operations were shuffled so much so that the local checkpoint was not advanced enough to allow us to reclaim documents or tombstone. Then I introduced this shuffle method.